modules/core/src/transitions/gpu-interpolation-transition.js

```diff
@@ -0,0 +1,225 @@
+import GL from '@luma.gl/constants';
+import {Buffer, Transform} from '@luma.gl/core';
+import {padArray} from '../utils/array-utils';
+import BaseAttribute from '../lib/base-attribute';
+import Attribute from '../lib/attribute';
+import Transition from './transition';
+
+export default class GPUInterpolationTransition {
+  constructor({gl, attribute, timeline}) {
+    this.transition = new Transition({timeline});
+    this.attribute = attribute;
+    // `attribute.userData` is the original options passed when constructing the attribute.
+    // This ensures that we set the proper `doublePrecision` flag and shader attributes.
+    this.attributeInTransition = new Attribute(gl, attribute.userData);
+    this.bufferLayout = attribute.bufferLayout;
+    // storing curLength because this.buffer may be larger than the actual length we want to use
+    // this is because we only reallocate buffers when they grow, not when they shrink,
+    // due to performance costs
+    this.curLength = null;
+    this.transform = null;
+    this.buffer = null;
+    this._swapBuffer = null;
+  }
+
+  isTransitioning() {
+    return Boolean(this.buffer);
+  }
+
+  getTransitioningAttribute() {
+    return this.attributeInTransition;
+  }
+
+  // Start a new transition using the current settings
+  // Updates transition state and from/to buffer
+  start(gl, settings, numInstances) {
+    const {fromState, toState, buffer} = this._setupBuffers(gl, settings, numInstances);
+    this.buffer = buffer;
+    const elementCount = Math.floor(this.curLength / this.attribute.size);
+    this.transition.start(settings);
+    if (this.transform) {
+      this.transform.update({
+        ...getBuffers({fromState, toState, buffer}),
+        elementCount
+      });
+    } else {
+      this.transform = new Transform(gl, {
+        elementCount,
+        ...getShaders(this.attribute.size),
+        ...getBuffers({fromState, toState, buffer})
+      });
+    }
+  }
+
+  update() {
+    const updated = this.transition.update();
+    if (updated) {
+      this.transform.run({
+        uniforms: {time: this.transition.time}
+      });
+    }
+    return updated;
+  }
+
+  cancel() {
+    this.transition.cancel();
+    this.transform.delete();
+    if (this.buffer) this.buffer.delete();
+    if (this._swapBuffer) this._swapBuffer.delete();
+  }
+
+  // get current values of an attribute, clipped/padded to the size of the new buffer
+  _setupBuffers(gl, settings, numInstances) {
+    const {size, normalized} = this.attribute;
```

Comment:
I would expect the `fromState` and `toState` preparation the same for all transition types...

Reply:
For physics-based transitions this function will be a little different because there are essentially two "from" buffers to update - the "current value" and the "previous value". But yes, there will be logic we can share between the transition types.

Reply:
Can we split off the common logic into a base class e.g. `GPUTransition`?

Reply:
Yeah - I think it might make sense to do that in the spring-transition work.